source/common/http/utility.cc

```diff
@@ -747,5 +749,74 @@ std::string Utility::PercentEncoding::decode(absl::string_view encoded) {
   return decoded;
 }
 
+const Utility::AuthorityAttributes Utility::parseAuthority(const absl::string_view& authority,
+                                                           uint32_t default_port) {
+  Utility::AuthorityAttributes auth_attr;
```

Comment:
This overall seems too much change for a refactoring from https://github.com/envoyproxy/envoy/blob/master/source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc#L93-L130, I think we can just copy that and adding `return {is_ip_address, host_to_resolve, default_port};` at the end as a start point.

That probably doesn't solves the issue by @htuch but less churn and unblock #8889. Then we can add bracket handling into `parseInternetAddress` later. WDYT @htuch?

Reply:
Can we write down a 3-4 step pseudo-code for what this is intended to do? I think the idea is it's trying to guess if it's an IP address with port, then maybe IP then host name? I don't see the connection between port and IP address implied below.

I would rather we do something clean and deliberative here, that is as simple and reliable as possible. Ideally we are following some RFC recommended approach to extracting this information to align us with other systems that will also interpret these strings.

Reply:
@htuch Pseudo-code is separated four sections to parse passed authority correctly.

```
if authority has port
	try to parse authority with Network::Utility::parseInternetAddressAndPort
	if succeeded to try parsing // section 1 (Authority should be IP address with port)
		determined authority is pure IP address and that is not FQDN
                is_ip_address ← true
		return result
	else // section 2 (Authority have IP address but parsing should be fail because authority have FQDN)
		// maybe passed authority is FQDN with port
		port ← Network::Utility::portFromIpAddress
		host ← Network::Utility::hostFromIpAddress
		is_ip_address ← false // that is because of failed to parse above function
		return result
	end if
end if

host ← Network::Utility::hostFromIpAddress
try to parse authority with Network::Utility::parseInternetAddress
If succeeded to try parsing // section 3 (Authority should not have port, maybe IP address only host)
	determined authority is pure IP address that have only host
        is_ip_address ← true
	return result
else // section 4 (Authority should be pure FQDN or FQDN with port)
	determined authority is not pure IP address because that is FQDN
        is_ip_address ← false
	return result
end if
```

Target authorities are like

- (IPv4) 123.456.789.000:8000
- (IPv6) [::]:8000

These should be not FQDN and having host and port. Parsed correctly on section 1 in pseudo-code

- (FQDN) example.com:8000

These should be FQDN and having port. Parsed correctly on section 2 in pseudo-code

- (IPv4) 123.456.789.000
- (IPv6) [::]
- (IPv6) fd12:3456:7890:1244:5678:9012:3456:7890

These should not be FQDN and that is host only IP address. Parsed correctly on section 3 in pseudo-code

- (FQDN) example.com
- (FQDN) example.com:abc

These should be FQDN and not having port. Parsed correctly on section 4 in pseudo-code

Reply:
Thanks for the overview. I think this should be distilled to comments. @PiotrSikora are there any relevant RFC algorithms for this we should look at?

Reply:
I think https://tools.ietf.org/html/rfc3986#section-3.2.2 is the relevant RFC. I think what you have above matches, i.e. first-match-wins, but it would be good to reference the RFC and explain how this maps to your code. You don't need to do pseudo-code, but code comments should explain how what you are doing is matching what the RFC says.